File: typescript/day10_part2_2019.ts
interface Asteroid {
    x: number;
    y: number;
    angle: number;
    dist: number;
}

function main() {
    const asteroids = readAsteroids("input.txt");
    const station = findBestAsteroidLocation(asteroids);
    const vaporized = vaporizeAsteroids(asteroids, station);
    if (vaporized.length >= 200) {
        const result = vaporized[199].x * 100 + vaporized[199].y;
        console.log(result);
    } else {
        console.log("Less than 200 asteroids were vaporized.");
    }
}

function readAsteroids(filename: string): boolean[][] {
    const fs = require('fs');
    const lines = fs.readFileSync(filename, 'utf-8').split('\n');
    return lines.map((line: string) => Array.from(line).map((char: string) => char === '#'));
}

function vaporizeAsteroids(asteroids: boolean[][], station: [number, number]): Asteroid[] {
    const targets: Asteroid[] = [];
    for (let y = 0; y < asteroids.length; y++) {
        for (let x = 0; x < asteroids[y].length; x++) {
            if (asteroids[y][x] && !(x === station[0] && y === station[1])) {
                const angle = Math.atan2(y - station[1], x - station[0]);
                const dist = Math.hypot(x - station[0], y - station[1]);
                const adjustedAngle = angle < -Math.PI / 2 ? angle + 2 * Math.PI : angle;
                targets.push({ x, y, angle: adjustedAngle, dist });
            }
        }
    }

    targets.sort((a, b) => a.angle === b.angle ? a.dist - b.dist : a.angle - b.angle);

    const vaporized: Asteroid[] = [];
    while (targets.length > 0) {
        let lastAngle = -Infinity;
        for (let i = 0; i < targets.length;) {
            if (targets[i].angle !== lastAngle) {
                vaporized.push(targets[i]);
                lastAngle = targets[i].angle;
                targets.splice(i, 1);
            } else {
                i++;
            }
        }
    }
    return vaporized;
}

function findBestAsteroidLocation(asteroids: boolean[][]): [number, number] {
    let maxCount = 0;
    let bestLocation: [number, number] = [0, 0];
    for (let y = 0; y < asteroids.length; y++) {
        for (let x = 0; x < asteroids[y].length; x++) {
            if (asteroids[y][x]) {
                const count = countVisibleAsteroids(asteroids, x, y);
                if (count > maxCount) {
                    maxCount = count;
                    bestLocation = [x, y];
                }
            }
        }
    }
    return bestLocation;
}

function countVisibleAsteroids(asteroids: boolean[][], x: number, y: number): number {
    const angles = new Set<number>();
    for (let otherY = 0; otherY < asteroids.length; otherY++) {
        for (let otherX = 0; otherX < asteroids[otherY].length; otherX++) {
            if (asteroids[otherY][otherX] && !(otherX === x && otherY === y)) {
                const angle = Math.atan2(otherY - y, otherX - x);
                angles.add(angle);
            }
        }
    }
    return angles.size;
}

main();

File: swift/day10_part2_2019.swift

import Foundation

struct Asteroid: Comparable {
    let x: Int
    let y: Int
    let angle: Double
    let dist: Double

    static func < (lhs: Asteroid, rhs: Asteroid) -> Bool {
        if lhs.angle == rhs.angle {
            return lhs.dist < rhs.dist
        }
        return lhs.angle < rhs.angle
    }
}

func readAsteroids(filename: String) -> [[Bool]] {
    guard let fileURL = URL(string: "file://\(FileManager.default.currentDirectoryPath)/\(filename)"),
          let fileContents = try? String(contentsOf: fileURL) else {
        fatalError("Failed to read file")
    }
    return fileContents.components(separatedBy: "\n").filter{!$0.isEmpty}.map { line in
        line.map { $0 == "#" }
    }
}

func vaporizeAsteroids(asteroids: [[Bool]], station: (x: Int, y: Int)) -> [Asteroid] {
    var targets: [Asteroid] = []
    for y in asteroids.indices {
        for x in asteroids[y].indices {
            if asteroids[y][x] && !(x == station.x && y == station.y) {
                var angle = atan2(Double(y - station.y), Double(x - station.x))
                let dist = hypot(Double(x - station.x), Double(y - station.y))
                if angle < -Double.pi / 2 {
                    angle += 2 * Double.pi
                }
                targets.append(Asteroid(x: x, y: y, angle: angle, dist: dist))
            }
        }
    }
    targets.sort()
    var vaporized: [Asteroid] = []
    var lastAngle = -Double.infinity
    while !targets.isEmpty {
        var i = 0
        while i < targets.count {
            if targets[i].angle != lastAngle {
                vaporized.append(targets[i])
                lastAngle = targets[i].angle
                targets.remove(at: i)
            } else {
                i += 1
            }
        }
        lastAngle = -Double.infinity
    }
    return vaporized
}

func findBestAsteroidLocation(asteroids: [[Bool]]) -> ((x: Int, y: Int), maxCount: Int) {
    var bestLocation = (x: 0, y: 0)
    var maxCount = 0
    for y in asteroids.indices {
        for x in asteroids[y].indices {
            if asteroids[y][x] {
                let count = countVisibleAsteroids(asteroids: asteroids, x: x, y: y)
                if count > maxCount {
                    maxCount = count
                    bestLocation = (x: x, y: y)
                }
            }
        }
    }
    return (bestLocation, maxCount)
}

func countVisibleAsteroids(asteroids: [[Bool]], x: Int, y: Int) -> Int {
    var angles: Set<Double> = []
    for otherY in asteroids.indices {
        for otherX in asteroids[otherY].indices {
            if asteroids[otherY][otherX] && !(otherX == x && otherY == y) {
                let angle = atan2(Double(otherY - y), Double(otherX - x))
                angles.insert(angle)
            }
        }
    }
    return angles.count
}

let asteroids = readAsteroids(filename: "input.txt")
let (station, _) = findBestAsteroidLocation(asteroids: asteroids)
let vaporized = vaporizeAsteroids(asteroids: asteroids, station: station)

if vaporized.count >= 200 {
    let result = vaporized[199].x * 100 + vaporized[199].y
    print(result)
} else {
    print("Less than 200 asteroids were vaporized.")
}
